File: lib/sudokuSolver.mjs
"use strict";

/**************************
 * Import important stuff *
 **************************/

import _ from "lodash";

/***********
 * Helpers *
 ***********/

/**
 * Extracts the 'value' field from an object. For use as a mapping function
 *
 * @private
 */
function extractValue({value}) {return value;}

/*****************************
 * The sudokuSolver function *
 *****************************/

/**
 * Solves a sudoku by recursive brute force
 *
 * @param {SudokuBoard} board	The sudoku to solve
 * @param {Integer} [startIndex]	The index to start looking for empty cells on. Defaults to 0, i.e. the first cell
 *
 * @returns {SudokuBoard}	The solved version of the input board
 *
 * @throws {Error}	If the board is unsolvable
 */
function sudokuSolver(board, startIndex = 0) {
	// Check if the board is solved
	if (board.isSolved()) {
		return board;
	} else if (startIndex === 0 && board.getErrors().length > 0) {
		throw new Error("Board is unsolvable");
	}

	// Find the first empty cell
	let cellIndex = startIndex;
	while (board.getCell(cellIndex).value !== 0) {
		cellIndex++;
	}

	// Handle for the solved board
	let solvedBoard = null;

	// Find out what values the cell can possibly have
	let cell = board.getCell(cellIndex);
	let possibleValues = _.difference([1, 2, 3, 4, 5, 6, 7, 8, 9],
		_.map(board.getRow(cell.row), extractValue),
		_.map(board.getCol(cell.col), extractValue),
		_.map(board.getSquare(cell.square), extractValue)
	);

	while (possibleValues.length > 0 && solvedBoard === null) {
		// Set the cell to the next value
		board.setCellValue(cellIndex, possibleValues.pop());

		try {
			// Recurse down to solve next cell
			solvedBoard = sudokuSolver(board.clone(), cellIndex+1);
		} catch (err) {
			// There were no solutions with this number. Do nothing
		}
	}

	// Check if the board was actually solved
	if (solvedBoard === null) {
		throw new Error("Board is unsolvable");
	}

	// The board is now solved!
	return solvedBoard;
}

/*********************
 * Export the solver *
 *********************/

export default sudokuSolver;
export { sudokuSolver };

File: Python3/sudokuSolver.py
##########################
# Import important stuff #
##########################

from SudokuBoard import SudokuBoard

############################
# Define the Sudoku solver #
############################

def sudoku_solver(board, start_index = 0):
    """Solves a SudokuBoard"""

    # Check if the board is solved
    if board.is_solved():
        return board
    elif start_index == 0 and len(board.get_errors()) > 0:
        raise Exception("Board is unsolvable")

    # Find the first empty cell
    cell_index = start_index
    while board.get_cell(cell_index)["value"] != SudokuBoard.EMPTY_VALUE:
        cell_index += 1

    # Handle for the solved board
    solved_board = None

    # Find out what values the cell can possibly have
    cell = board.get_cell(cell_index)
    possible_values = set(range(1, 10))
    possible_values = possible_values - set([cell["value"] for cell in board.get_row(cell["row"])])
    possible_values = possible_values - set([cell["value"] for cell in board.get_col(cell["col"])])
    possible_values = possible_values - set([cell["value"] for cell in board.get_square(cell["square"])])
    possible_values = list(possible_values)

    # Try all possible values until a solution is found
    while len(possible_values) > 0 and solved_board == None:
        # Set the cell to the next possible value
        board.set_cell_value(cell_index, possible_values.pop())

        try:
            # Recurse down to solve the next cell
            solved_board = sudoku_solver(board.clone(), cell_index+1)
        except:
            # There were no solutions. Don't do anything
            pass


    # Check if the board was solved
    if solved_board == None:
        # Reset the cell
        board.set_cell_value(cell_index, SudokuBoard.EMPTY_VALUE)

        # Complain loudly
        raise Exception("Board is unsolvable")

    # The board is now solved!
    return solved_board
